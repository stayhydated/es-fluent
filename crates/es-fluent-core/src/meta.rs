use strum::{Display, EnumString};

pub struct StructKind;
pub struct EnumKind;

#[derive(Clone, Debug, Display, EnumString, Eq, Hash, PartialEq, serde::Serialize)]
pub enum TypeKind {
    Struct,
    Enum,
}
