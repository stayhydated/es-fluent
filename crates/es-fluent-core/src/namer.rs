use derive_more::{Debug, Deref, Display};
use heck::ToSnakeCase as _;
use quote::format_ident;

#[derive(Clone, Debug, Deref, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct FluentKey(pub String);

impl FluentKey {
    pub const DELIMITER: &str = "-";
    pub fn new(ftl_name: &syn::Ident, sub_name: &str) -> Self {
        let normalized_name = ftl_name.to_string().to_snake_case();
        if sub_name.is_empty() {
            FluentKey(normalized_name)
        } else {
            FluentKey(format!(
                "{}{}{}",
                normalized_name,
                Self::DELIMITER,
                sub_name
            ))
        }
    }
}

impl quote::ToTokens for FluentKey {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let key_string = &self.0;
        tokens.extend(quote::quote! { #key_string });
    }
}

#[derive(Clone, Debug, Deref, Display, Eq, Hash, PartialEq)]
pub struct FluentDoc(pub String);

impl From<&FluentKey> for FluentDoc {
    fn from(ftl_key: &FluentKey) -> Self {
        FluentDoc(format!("Key = `{}`", *ftl_key))
    }
}

impl quote::ToTokens for FluentDoc {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let doc_string = &self.0;
        tokens.extend(quote::quote! { #doc_string });
    }
}

pub struct UnnamedItem(usize);

impl std::fmt::Display for UnnamedItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", Self::UNNAMED_PREFIX, self.0)
    }
}

impl UnnamedItem {
    const UNNAMED_PREFIX: &str = "f";

    pub fn to_ident(&self) -> syn::Ident {
        format_ident!("{}", self.to_string())
    }
}

impl From<usize> for UnnamedItem {
    fn from(index: usize) -> Self {
        Self(index)
    }
}
