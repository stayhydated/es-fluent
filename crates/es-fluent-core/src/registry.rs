use crate::meta::TypeKind;
use crate::namer::FluentKey;
use bon::Builder;

#[derive(Builder, Clone, Debug, Eq, Hash, PartialEq)]
pub struct FtlVariant {
    pub name: String,
    pub ftl_key: FluentKey,
    pub arguments: Option<Vec<String>>,
}

#[derive(Builder, Clone, Debug, Eq, Hash, PartialEq)]
pub struct FtlTypeInfo {
    pub type_kind: TypeKind,
    pub type_name: String,
    pub variants: Vec<FtlVariant>,
}
