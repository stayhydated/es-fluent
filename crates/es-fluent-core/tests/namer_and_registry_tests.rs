use es_fluent_core::meta::TypeKind;
use es_fluent_core::namer::{FluentKey, UnnamedItem};
use es_fluent_core::registry::{FtlTypeInfo, FtlVariant};
use proc_macro2::Span;
use syn::Ident;

#[test]
fn fluent_key_new_with_and_without_subname() {
    let ident = Ident::new("TestName", Span::call_site());

    let key_with_sub = FluentKey::new(&ident, "sub_part");
    assert_eq!(key_with_sub.to_string(), "test_name-sub_part");

    let key_without_sub = FluentKey::new(&ident, "");
    assert_eq!(key_without_sub.to_string(), "test_name");

    let same_key = FluentKey::new(&ident, "");
    assert_eq!(key_without_sub, same_key);
    let different_key = FluentKey::new(&ident, "x");
    assert_ne!(key_without_sub, different_key);
}

#[test]
fn unnamed_item_display_and_ident() {
    let item_zero = UnnamedItem::from(0usize);
    assert_eq!(item_zero.to_string(), "f0");
    assert_eq!(item_zero.to_ident().to_string(), "f0");

    let item_five: UnnamedItem = 5usize.into();
    assert_eq!(item_five.to_string(), "f5");
    assert_eq!(item_five.to_ident().to_string(), "f5");

    assert_ne!(item_zero.to_string(), item_five.to_string());
}

#[test]
fn registry_structs_equality() {
    let ident = Ident::new("ErrorType", Span::call_site());

    let v1 = FtlVariant::builder()
        .name("Network".to_string())
        .ftl_key(FluentKey::new(&ident, "Network"))
        .maybe_arguments(Some(vec!["code".into(), "message".into()]))
        .build();

    let v1_dup = FtlVariant::builder()
        .name("Network".to_string())
        .ftl_key(FluentKey::new(&ident, "Network"))
        .maybe_arguments(Some(vec!["code".into(), "message".into()]))
        .build();

    let v2 = FtlVariant::builder()
        .name("Io".to_string())
        .ftl_key(FluentKey::new(&ident, "Io"))
        .maybe_arguments(None)
        .build();

    assert_eq!(v1, v1_dup);
    assert_ne!(v1, v2);

    let info = FtlTypeInfo::builder()
        .type_kind(TypeKind::Struct)
        .type_name("ErrorType".to_string())
        .variants(vec![v1.clone(), v2.clone()])
        .build();

    let info_dup = FtlTypeInfo::builder()
        .type_kind(TypeKind::Struct)
        .type_name("ErrorType".to_string())
        .variants(vec![v1, v2])
        .build();

    assert_eq!(info, info_dup);
}

#[test]
fn snapshot_ftl_variant_and_typeinfo_debug() {
    let ident = Ident::new("ErrorType", Span::call_site());

    let variant = FtlVariant::builder()
        .name("Network".to_string())
        .ftl_key(FluentKey::new(&ident, "Network"))
        .maybe_arguments(Some(vec!["code".into(), "message".into()]))
        .build();

    insta::assert_ron_snapshot!("snapshot_ftl_variant_and_typeinfo_debug__variant", &variant);

    let variant2 = FtlVariant::builder()
        .name("Io".to_string())
        .ftl_key(FluentKey::new(&ident, "Io"))
        .maybe_arguments(None)
        .build();

    let type_info = FtlTypeInfo::builder()
        .type_kind(TypeKind::Struct)
        .type_name("ErrorType".to_string())
        .variants(vec![variant.clone(), variant2.clone()])
        .build();

    insta::assert_ron_snapshot!(
        "snapshot_ftl_variant_and_typeinfo_debug__type_info",
        &type_info
    );
}
