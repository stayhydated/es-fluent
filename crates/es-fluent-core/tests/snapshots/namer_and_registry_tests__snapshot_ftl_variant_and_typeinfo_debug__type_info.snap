---
source: crates/es-fluent-core/tests/namer_and_registry_tests.rs
expression: "&type_info"
---
FtlTypeInfo(
  type_kind: Struct,
  type_name: "ErrorType",
  variants: [
    FtlVariant(
      name: "Network",
      ftl_key: FluentKey("error_type-Network"),
      arguments: Some([
        "code",
        "message",
      ]),
    ),
    FtlVariant(
      name: "Io",
      ftl_key: FluentKey("error_type-Io"),
      arguments: None,
    ),
  ],
)
